File: 1569-max-dot-product-of-two-subsequences/max-dot-product-of-two-subsequences.py
class Solution:
    def maxDotProduct(self, nums1: List[int], nums2: List[int]) -> int:
        m, n = len(nums1), len(nums2)

        dp = [[float('-inf')] * (n + 1) for _ in range(m + 1)]

        for i in range(1, m + 1):
            for j in range(1, n + 1):
                curr = nums1[i - 1] * nums2[j - 1]
                dp[i][j] = max(dp[i - 1][j - 1], curr,  dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1] + curr)
                
        return dp[-1][-1]

File: 1569-max-dot-product-of-two-subsequences/max-dot-product-of-two-subsequences.cpp
class Solution {
public:
    int maxDotProduct(std::vector<int>& nums1, std::vector<int>& nums2) {
        int m = nums1.size(), n = nums2.size();
        vector<int> dp(n + 1, INT_MIN), pre(n + 1, INT_MIN);

        for (int i = 1; i <= m; ++i) {
            for (int j = 1; j <= n; ++j) {
                int curr = nums1[i - 1] * nums2[j - 1];
                dp[j] = max({curr, pre[j], dp[j - 1], curr + max(0, pre[j - 1])});
            }
            swap(dp, pre);
        }
        
        return pre[n];
    }
};